server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java

```diff
@@ -117,7 +147,44 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
                     "[span_multi] must have [" + MATCH_FIELD.getPreferredName() + "] multi term query clause");
         }
 
-        return new SpanMultiTermQueryBuilder(subQuery).queryName(queryName).boost(boost);
+        return new SpanMultiTermQueryBuilder(subQuery).queryName(queryName).boost(boost).maxExpansions(maxExpansionValue);
+    }
+
+    public class TopTermSpanBooleanQueryRewriteWithMaxClause extends SpanMultiTermQueryWrapper.SpanRewriteMethod{
+        MultiTermQuery multiTermQuery;
+
+        @Override
+        public SpanQuery rewrite(IndexReader reader, MultiTermQuery query) throws IOException {
+            multiTermQuery = query;
+            return (SpanQuery) this.delegate.rewrite(reader, multiTermQuery);
+        }
+
+        final ScoringRewrite<List<SpanQuery>> delegate = new ScoringRewrite<List<SpanQuery>>() {
+
+            @Override
+            protected List<SpanQuery> getTopLevelBuilder() {
+                return new ArrayList();
+            }
+
+            @Override
+            protected Query build(List<SpanQuery> builder) {
+                return new SpanOrQuery((SpanQuery[]) builder.toArray(new SpanQuery[builder.size()]));
+            }
+
+            @Override
+            protected void checkMaxClauseCount(int count) {
+                if (count > maxExpansions) {
+                    throw new ElasticsearchException("[" + multiTermQuery.toString() + " ] " +
+                        "exceeds maxExpansions [maxExpansions is set to " + maxExpansions + "]");
+                }
+            }
+
+            @Override
+            protected void addClause(List<SpanQuery> topLevel, Term term, int docCount, float boost, TermContext states) {
+                SpanTermQuery q = new SpanTermQuery(term, states);
+                topLevel.add(q);
+            }
+        };
```

Comment:
I would not throw an error, I would just ensure that we collect only up to `maxExpansions` queries. @jimczi what do you think?

Reply:
Thanks , I will look at checklist . I thought the plan was to throw exception if term expansion count is above max_expansions ( which is defaulted to 1024 ) . The other pattern can be done already by adding ```rewrite``` param to inner multi query

Reply:
I think it's better to throw an exception, at least in master. I don't think we need a `max_expansion` option though, we can just honor `BooleanQuery#getMaxClauseCount()` and fails the query if there are more terms. The span scoring rewrite does not allow to limit the number of expansions, it is all or nothing and I think it's a good property since it will not confuse the users by returning partial responses on prefix queries that match more than 1024 terms. This is the main reason why I opened the issue in the first place, we should never build a disjunction on more than `BooleanQuery.getMaxClauseCount()`.